File: javascript/triangle/triangle.js
export class Triangle {

    #sideA;
    #sideB;
    #sideC;

    constructor(sideA, sideB, sideC) {
        this.#sideA = sideA;
        this.#sideB = sideB;
        this.#sideC = sideC;
    }

    get isEquilateral() {
        return this.#isTriangle() && this.#sideA === this.#sideB 
                                && this.#sideB === this.#sideC;
    }

    get isIsosceles() {
        return this.#isTriangle() && (this.#sideA === this.#sideB 
                || this.#sideB === this.#sideC || this.#sideA === this.#sideC);
    }

    get isScalene() {
        return this.#isTriangle() 
                && (this.#sideA !== this.#sideB 
                    && this.#sideA !== this.#sideC 
                    && this.#sideB !== this.#sideC);
    }

    #isTriangle() {
        return this.#sideA > 0 && this.#sideB > 0 && this.#sideC > 0
                    && ((this.#sideA + this.#sideB >= this.#sideC) &&
                        (this.#sideB + this.#sideC >= this.#sideA) &&
                        (this.#sideA + this.#sideC >= this.#sideB)); 
    }
}

File: kotlin/triangle/src/main/kotlin/Triangle.kt
class Triangle<out T : Number> (val a: T, val b: T, val c: T) {

    init {

        val dA: Double = a.toDouble()
        val dB: Double = b.toDouble()
        val dC: Double = c.toDouble()

        require (dA + dB > dC)
        require (dA + dC > dB)
        require (dB + dC > dA)
        require (dA > 0 && dB > 0 && dC > 0)

    }

    val isEquilateral: Boolean = a == b && b == c
    val isIsosceles: Boolean = a == b || a == c || b == c
    val isScalene: Boolean = a != b && a != c && b != c


}
